prometheus/registry_test.go

```diff
@@ -746,38 +746,153 @@ func BenchmarkHandler(b *testing.B) {
 }
 
 func TestAlreadyRegistered(t *testing.T) {
-	reg := prometheus.NewRegistry()
 	original := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name:        "test",
+			Help:        "help",
+			ConstLabels: prometheus.Labels{"const": "label"},
+		},
+		[]string{"foo", "bar"},
+	)
+	equalButNotSame := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name:        "test",
+			Help:        "help",
+			ConstLabels: prometheus.Labels{"const": "label"},
+		},
+		[]string{"foo", "bar"},
+	)
+	originalWithoutConstLabel := prometheus.NewCounterVec(
 		prometheus.CounterOpts{
 			Name: "test",
 			Help: "help",
 		},
 		[]string{"foo", "bar"},
 	)
-	equalButNotSame := prometheus.NewCounterVec(
+	equalButNotSameWithoutConstLabel := prometheus.NewCounterVec(
 		prometheus.CounterOpts{
 			Name: "test",
 			Help: "help",
 		},
 		[]string{"foo", "bar"},
 	)
 	var err error
-	if err = reg.Register(original); err != nil {
-		t.Fatal(err)
-	}
-	if err = reg.Register(equalButNotSame); err == nil {
-		t.Fatal("expected error when registering equal collector")
-	}
-	if are, ok := err.(prometheus.AlreadyRegisteredError); ok {
-		if are.ExistingCollector != original {
-			t.Error("expected original collector but got something else")
+
+	t.Run("RegisterNormallyReregisterNormally", func(t *testing.T) {
+		reg := prometheus.NewRegistry()
+		if err = reg.Register(original); err != nil {
+			t.Fatal(err)
 		}
-		if are.ExistingCollector == equalButNotSame {
-			t.Error("expected original callector but got new one")
+
+		if err = reg.Register(equalButNotSame); err == nil {
+			t.Fatal("expected error when registering equal collector")
 		}
-	} else {
-		t.Error("unexpected error:", err)
-	}
+		if are, ok := err.(prometheus.AlreadyRegisteredError); ok {
+			if are.ExistingCollector != original {
+				t.Error("expected original collector but got something else")
+			}
+			if are.ExistingCollector == equalButNotSame {
```

Comment:
I think I would just have that included in the error above
```go
t.Errorf("expected original collector but got %v", are.ExiststingCollector)
```
(or even use Fatalf)

Reply:
The %v representation of a collector can look a bit daunting. As this is an `==` test, you'll see from the line where the test fails what precise collector was returned. My take is that this is slightly clearer than dumping a possibly complex text representation to the test output (which might have the right hint for the qualified reader, but will probably require to look at the test code anyway).